convert2rhel/subscription.py

```diff
@@ -277,28 +297,144 @@ def args(self):
+            path = urllib.parse.urlunsplit(("", "", url_parts.path, url_parts.query, url_parts.fragment))
+            if path:
+                connection_opts["handler"] = path
+
+        connection_opts = dbus.Dictionary(connection_opts, signature="ss")
+        return connection_opts
+
+    def __call__(self):
+        """
+        Use dbus to register the system with subscription-manager.
+
+        Status of dbus on various platforms:
+            * RHEL6:
+                * DBUS-1.2.24 is present but may not be installed.
+                * Install the dbus package and run /etc/rc.d/init.d/messagebus start
+                * dbus-python-0.83.0 is available
+            * RHEL7:
+                * dbus-1.10.24 is installed and run by default
+                * dbus-python-1.1.9 is available
+            * RHEL8 & RHEL9
+                * dbus-1.12.x is installed and run by default
+                * python3-dbus-1.2.x is available
+
+        .. seealso::
+            Documentation for the subscription-manager dbus API:
+            https://www.candlepinproject.org/docs/subscription-manager/dbus_objects.html
+        """
+        # DBus demands that the full signature of a dictionary (what the datatypes of the keys and
+        # values will be) is known ahead of time.  An empty python dict doesn't have enough
+        # information and dbus-python doesn't like {"": ""} even though /bin/dbus-send thinks that
+        # is okay.  Use this instead.
+        EMPTY_DBUS_DICT = dbus.Dictionary({}, signature="ss")
+
+        loggerinst.debug("Getting a handle to the system dbus")
+        system_bus = dbus.SystemBus()
+
+        # Create a new bus so we can talk to rhsm privately (For security:
+        # talking on the system bus might be eavesdropped in certain scenarios)
+        loggerinst.debug("Getting a subscription-manager RegisterServer object from dbus")
+        register_server = system_bus.get_object("com.redhat.RHSM1", "/com/redhat/RHSM1/RegisterServer")
+        loggerinst.debug("Starting a private DBus to talk to subscription-manager")
+        address = register_server.Start(
+            i18n.SUBSCRIPTION_MANAGER_LOCALE, dbus_interface="com.redhat.RHSM1.RegisterServer"
+        )
+
+        try:
+            # Use the private bus to register the machine
+            loggerinst.debug("Connecting to the private DBus")
+            private_bus = dbus.connection.Connection(address)
+            loggerinst.debug("Getting a subscription-manager Register object from dbus")
+            register_object = private_bus.get_object("com.redhat.RHSM1", "/com/redhat/RHSM1/Register")
+
+            if self.password:
+                loggerinst.debug("Registering via username/password: %s" % " ".join(hide_secrets(self.args)))
+                register_object.Register(
+                    self.org or "",
+                    self.username,
+                    self.password,
+                    EMPTY_DBUS_DICT,
```

Comment:
Please replace the empty dict with the options dict enabling the `force` option.

Reply:
This has now been fixed in the PR.